_posts/2020-10-27-scalajs-for-scala-3.md

```diff
@@ -0,0 +1,184 @@
+
+There are many components that contribute to this behavior.
+Just looking at the compilation pipeline, there are already 4 components involved: the compiler plugin for scalac, the linker, and the artifacts for the Scala standard library and the JDK subset that is implemented.
+
+### A compiler plugin
+
+When compiling a Scala.js codebase, the .scala source files are compiled with scalac, augmented with the Scala.js compiler plugin.
+From scalac, we receive parsing, type checking, elaboration (type inference, implicit resolution, overload resolution, etc.), as well as many transformation phases.
+The compiler plugin, comparatively to the rest of scalac, is small: it takes the internal compiler ASTs that have been lowered to contain JVM-style classes, interfaces, methods and instructions, and turns them into Scala.js IR (.sjsir files).
+
+Thanks to everything it receives from scalac for free, the Scala.js compiler plugin does not have to deal with a lot of Scala-specific features.
+Traits have been lowered to interfaces; lazy vals have been desugared; nested classes, defs, and functions have been translated to top-level classes with methods; etc.
+This is how Scala.js is able to keep up-to-date with all the versions of Scala 2, and be available on the same day as new Scala releases.
+Not all compile-to-JS languages have that property!
+
+### `.sjsir`files
+
+.sjsir files are similar to .class files, although they are AST-based (instead of stack-machine-based) and contain features dedicated to JavaScript interoperability.
+The .sjsir format and specification is independent of Scala, its compiler or its standard library.
+In fact, there is nothing Scala-specific in the entire linker, other than a few optimizations targeted at Scala-like code.
+This is important because it means that the linker is independent of the version of Scala: we use the same linker irrespective of the version of Scala that was used to create the .sjsir files.
+In theory, we could even compile other JVM languages like Kotlin or Java to the Scala.js IR, then use our linker to emit JavaScript code.
+While we have a proof-of-concept for Java, it is nowhere near usable, however.
+
+.sjsir files have binary compatibility guarantees similar to those of .class files.
+With a few exceptions related to JS interop annotations, if changes to a .class files are binary compatible, then the changes applied to the corresponding .sjsir files are also binary compatible.
+This allows Scala.js codebase to meaningfully use MiMa to check that they do not break their binary API.
+
+The Scala.js linker (optimizer) takes the .sjsir files from the application and from all the transitive dependencies.
+In particular, the .sjsir files that are built for the Scala standard library and for the JDK subset.
+The linker optimizes all of them together, then emits one .js file for the entire codebase.
```

Comment:
Maybe consider adding a diagram to make it easier to get a global picture?

Reply:
Oh, there was a diagram in my draft. I forgot to include it in the .md port 🤦‍♂️

I'll add it back tomorrow morning.

Reply:
Done.